Bug/63898 missing form label in work package comments rich text editor raises browser console error

akabiru opened the pull request:
https://community.openproject.org/work_packages/63898

Amend the text area label requirement introduced in https://github.com/opf/openproject/commit/6bf31400ed07e0ad1141160e6c2f8362a9683d28 to allow for visually hidden label as not all use cases fit this requirement- but **_there should always be a reference for screen-readers._**

As CKEditor re-renders the text area aria-label in every focus/blur event, we rely on the focusTracker to keep the aria-labelled-by up to date.

akabiru committed:
Missing Form Label in work package comments rich text editor raises browser console error

https://community.openproject.org/work_packages/63898

Amend the text area label requirement introduced in 6bf31400ed07e0ad1141160e6c2f8362a9683d28 to allow for an alternate
ARIA-label as not all use cases fit this requirement- but there should always be a reference for screenreaders.
`aria-labelled-by` OR `aria-label`

As CKEditor re-renders the text area aria-label in every focus/blur event, we rely on the focusTracker to keep the aria-label
updated

akabiru committed:
Ensure label click subscription is only registered once

myabc commented on frontend/src/app/shared/components/editor/components/ckeditor-augmented-textarea/ckeditor-augmented-textarea.component.ts:
```diff
@@ -288,23 +301,30 @@ export class CkeditorAugmentedTextareaComponent extends UntilDestroyedMixin impl
+    } else if (this.textAreaAriaLabel) {
+      ckContent.setAttribute('aria-label', this.textAreaAriaLabel);
+    } else {
+      throw new Error(`Please provide a label for the textarea with id ${this.textAreaId} or set the textAreaAriaLabel property.`);
```
I think we should reserve throwing Errors for critical failures, and instead go for `console.error` or `console.warn` (thus, not halting execution):

```suggestion
      console.error(`Please provide a label for the textarea with id ${this.textAreaId} or set the textAreaAriaLabel property.`);
```

(although I'd love a second opinion on this comment - @HDinger or @oliverguenther?)

myabc commented on lib/primer/open_project/forms/rich_text_area.html.erb:
```diff
@@ -1,11 +1,13 @@
-    <%= builder.text_area(@input.name,
-                          id: @text_area_id,
-                          **@input.input_arguments) %>
+    <%= builder.text_area(
+          @input.name,
+          id: @text_area_id,
+          **@input.input_arguments
+        ) %>
```
nitpick: It'd be good to have whitespace and formatting (erb_lint, rubocop) fixes as separate commits - to make it easier to focus on the salient changes.

akabiru replied:
Sure, this erb lint fix is a dedicated commit: https://github.com/opf/openproject/pull/18903/commits/0c44eb6562eee7ed5133dd67b270979247e93046 unless you mean as a separate PR?

akabiru replied:
I'll split this out- as I'm no longer "touching" this file + will look into component unit tests in the same PR!

myabc commented on frontend/src/app/shared/components/editor/components/ckeditor-augmented-textarea/ckeditor-augmented-textarea.component.ts:
```diff
@@ -215,7 +218,17 @@ export class CkeditorAugmentedTextareaComponent extends UntilDestroyedMixin impl
+      if (isFocused) {
+        this.setLabel();
+      }
+    });
```
👍🏻 💯 for looking into this + documenting!

myabc commented on spec/features/activities/work_package/activity_tab_comment_editor_spec.rb:
```diff
@@ -118,6 +118,25 @@ def expect_editor_to_be_dismissed_with_confirmation(&)
+
+      activity_tab.expect_focus_on_editor
+
+      within_test_selector("augmented-text-area-notes") do
+        expect(page).to have_css(".ck-content[aria-label='Add a comment. Type @ to notify people.']")
+      end
```
You should be able to use [Capybara accessible selectors](https://github.com/citizensadvice/capybara_accessible_selectors) here. In fact, my original motivation for the change in introduced in 6bf3140 (PR #18478) was to fix labels so that they would work with accessible selectors.

Either using the locator (see [`rich_text` locator](https://github.com/citizensadvice/capybara_accessible_selectors?tab=readme-ov-file#rich_text)):

```suggestion
      expect(page).to have_selector :rich_text, "Add a comment. Type @ to notify people."
```
OR without locator, but filtered by `aria-label` attribute (see [`aria` [Hash] docs](https://github.com/citizensadvice/capybara_accessible_selectors?tab=readme-ov-file#aria-hash)) - if you want to be explicit about the need for `aria-label` being set

```suggestion
      expect(page).to have_selector :rich_text, aria: { label: "Add a comment. Type @ to notify people." }
```

I'm personally not a big fan of using test selectors in Capybara feature specs - I prefer to drive specs using semantic elements and labels. If you do want to use test selectors in this scenario though, you could also write:

```suggestion
      within_test_selector("augmented-text-area-notes") do
        expect(page).to have_css ".ck-content", aria: { label: "Add a comment. Type @ to notify people." }
      end
```

akabiru replied:
Nice one, I'll try this out! 👍🏾

myabc commented on lib/primer/open_project/forms/rich_text_area.rb:
```diff
@@ -17,6 +17,14 @@ def initialize(input:, rich_text_options:)
+
+        private
+
+        def rich_text_options
+          @rich_text_options.tap do |options|
+            options[:textAreaAriaLabel] = options.delete(:aria_label) if options.key?(:aria_label)
+          end
+        end
```
we need to process various `rich_text_options` before being passed as `inputs:` to `angular_component_tag`. Let's either do the Hash munging in the ERb template OR in `rich_text_area.rb`, but not in both.

e.g. 

```rb
          @rich_text_options = rich_text_options
          @text_area_id = rich_text_options.delete(:text_area_id) || builder.field_id(@input.name)
          @text_area_aria_label = rich_text_options.delete(:aria_label) if rich_text_options.key?(:aria_label)
```

akabiru replied:
Agree, considered this but didn't have better ideas at the time :)

myabc reviewed (changes requested):
Hi @akabiru 

The TS/JS code looks good! Thanks for documenting why we need the `focusTracker`.

I've commented on the Ruby code in various places. 

As a general point, rather than passing an `aria_label` option to `rich_text_options`, I would prefer it if we instead update `RichTextAreaInput` to properly support the `visually_hide_label` option - thereby making it consistent with other form inputs, e.g.:

```ruby
      notes_form.rich_text_area(
        classes: "ck-editor-primer-adjusted",
        name: :notes,
        label: I18n.t("activities.work_packages.activity_tab.label_type_to_comment"),
        visually_hide_label: true,
        rich_text_options: {
          showAttachments: false,
          resource:,
          editor_type: "constrained"
        }
      )
```

In addition to the feature specs, it would be great to see some specs for `RichTextArea`/`RichTextAreaInput` that test the correct options/inputs are passed to the Angular component. While we would only be testing server HTML rendering (no JS!), there are enough permutations of form input options to make this worth testing.

* I've started adding form specs under `spec/forms` but they're very much a work in progress - and my goal would be to test the form components under `lib/primer/open_project` themselves.

Get in touch if you have any questions about my comments. Cheers A!

akabiru reviewed (commented):
Thanks for the review @myabc great points, happy to address 💯

akabiru committed:
Use `visually_hide_label` option instead

Stay consistent with primer as pointed out in the following review!
https://github.com/opf/openproject/pull/18903#pullrequestreview-2833035563

akabiru committed:
Add notes form unit spec, assert attributes passed

akabiru commented:
> 
> In addition to the feature specs, it would be great to see some specs for `RichTextArea`/`RichTextAreaInput` that test the correct options/inputs are passed to the Angular component. While we would only be testing server HTML rendering (no JS!), there are enough permutations of form input options to make this worth testing.
> 
>     * I've started adding form specs under `spec/forms` but they're very much a work in progress - and my goal would be to test the form components under `lib/primer/open_project` themselves.

Do you have a PR up for this, happy to add it on there. I took a stab and got stumped at how to stub out the `:builder` and `:form`, and I wonder if you've already solved this. I think we can aim for that to go into dev.

cbliard commented on frontend/src/app/shared/components/editor/components/ckeditor-augmented-textarea/ckeditor-augmented-textarea.component.ts:
```diff
@@ -288,23 +297,28 @@ export class CkeditorAugmentedTextareaComponent extends UntilDestroyedMixin impl
-      labelId = uniqueId('label-');
-      label.setAttribute('id', labelId);
-    }
+    if (label) {
+      let labelId;
```
I may be nitpicking, but I prefer an early return in a guard clause rather than a long if.

Something like this:

```js
const label = document.querySelector<HTMLLabelElement>(`label[for=${this.textAreaId}]`);
if (!label) {
  console.error(`Please provide a label for the textarea with id ${this.textAreaId}`);
  return
}

...
```

That makes things clearer for me.

akabiru replied:
Sure, I like guard clauses too. Will update 👍🏾 fwiw, I had an else if case before which is now removed, which is why the "original" structure is this way.

myabc commented on spec/features/activities/work_package/activity_tab_comment_editor_spec.rb:
```diff
@@ -118,6 +118,34 @@ def expect_editor_to_be_dismissed_with_confirmation(&)
+      expect(page).to have_selector(:label, for: "journal_notes", class: "sr-only",
+                                            text: "Add a comment. Type @ to notify people.")
+      label_id = page.find('label[for="journal_notes"]')["id"]
+      expect(page).to have_selector(:rich_text, aria: { labelledby: label_id })
```
```suggestion
      expect(page).to have_selector(:rich_text, "Add a comment. Type @ to notify people.")
```

akabiru committed:
Use guard clause to break early

Co-authored-by: Christophe Bliard <c.bliard@openproject.com>

akabiru committed:
Use capybara rich text selector for label assertion

The capybara acessible :rich_text selector includes an `aria-{label,labelled-by}` location finder
https://github.com/citizensadvice/capybara_accessible_selectors/blob/b60eebb9673b20dd472afb3d724340a6f6132bcb/lib/capybara_accessible_selectors/selectors/rich_text.rb#L18

Co-authored-by: Alexander Coles <a.coles@openproject.com>
